[3.3.5] CastSpell unclusterfucking (that's a word now)

Treeston opened the pull request:
* CastSpell now always takes three arguments - target, spellId, and a struct containing extra arguments
* This struct (CastSpellExtraArgs, see Unit.h) serves as a conglomerate of every previous combination of the 20 billion different CastSpell overloads, all merged into one. Also, `CastCustomSpell`, which is thus dead now.
  * It has some handy utility constructors - check it out! All of these can be used to implicitly construct the ExtraArgs object, so 99% of the current `CastSpell` invocation patterns should still work as is, just with significantly cleaner backing.

Note: Doesn't currently build - a crapton of scripts still need to be refactored. Posting this here so @Shauren can bitch about anything he wants changed before I invest more work.

PS: Obviously, the only relevant changes are in `Unit.cpp` and `.h`, with some minor sense-making refactoring in `Spell.cpp` (and `.h`). Everything else is make-it-build-again refactoring, and can be ignored for review purposes.

Treeston committed:
Core/Spell: The giant CastSpell unclusterfucking (that's a word now) of this generation.

- CastSpell now always takes three arguments - target, spellId, and a struct containing extra arguments
- This struct (CastSpellExtraArgs, see Unit.h) serves as a conglomerate of every previous combination of the 20 billion different CastSpell overloads, all merged into one
  - It has some great utility constructors - check it out! All of these can be used to implicitly construct the ExtraArgs object.

tkrokli commented:
Multiple chained errors in travis-ci, you may want to take a a look.

Shauren commented on src/server/game/Entities/Unit/Unit.h:
```diff
@@ -339,6 +339,24 @@ enum TriggerCastFlags : uint32
+    ObjectGuid originalCaster = ObjectGuid::Empty;
+    struct : std::vector<std::pair<SpellValueMod, int32>>
+    {
+        void add(SpellValueMod mod, int32 val) { push_back({ mod, val }); }
```
I will want this gone in future - just stick to push_back

joschiwald commented on src/server/game/Entities/Unit/Unit.h:
```diff
@@ -339,6 +311,30 @@ enum TriggerCastFlags : uint32
+        auto end() const { return data.cend(); }
+
+        std::vector<std::pair<SpellValueMod, int32>> data;
+    } customValues;
```
can we change this to allow method chaining?
```c++
CastSpellExtraArgs()
    .setTriggered()
    .addSpellValue(SPELLVALUE_BASE_POINT0, value)
    .setOriginalCaster(caster)
```

Treeston replied:
meh, i'm not particularly fond of it. Starts to get unreadable very quickly, and there's little benefit over the existing syntax.

Shauren replied:
Why not?
But both of you, dont forget TC codestyle - functions and public struct/class data members are start with uppercase

Treeston replied:
> starts to get unreadable very quickly
The primary use case is putting something like @joschiwald posted directly in the invocation. This makes for some very unreadable code.

joschiwald replied:
first time i saw that method chaining syntax, it was also a bit bad readable to me, but now i like it

to support chaining does not mean you must use it, however you like

Treeston replied:
it's not a question of whether _i_ would use it, it's a question of whether we want code using it in the core.

Shauren replied:
Look at MailDraft uses, I think it works pretty well

Treeston commented:
Merge ready pending travis.

Treeston commented:
travis passed, @Shauren review please

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -13664,11 +13663,15 @@ void Player::ApplyEnchantment(Item* item, EnchantmentSlot slot, bool apply, bool
-                            else
-                                CastSpell(this, enchant_spell_id, true, item);
+                            {
+                                for (int i = 0; i < 3; ++i)
```
uint32

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -13664,11 +13663,15 @@ void Player::ApplyEnchantment(Item* item, EnchantmentSlot slot, bool apply, bool
-                                CastCustomSpell(this, enchant_spell_id, &basepoints, &basepoints, &basepoints, true, item);
-                            else
-                                CastSpell(this, enchant_spell_id, true, item);
+                            {
```
brackets not needed from what i can see

Shauren commented on src/server/game/Spells/Auras/SpellAuras.cpp:
```diff
@@ -1281,9 +1282,13 @@ void Aura::HandleAuraSpecificMods(AuraApplication const* aurApp, Unit* caster, b
-                        target->CastCustomSpell(74396, SPELLVALUE_AURA_STACK, sSpellMgr->AssertSpellInfo(74396)->StackAmount, (Unit*)nullptr, true);
+                        CastSpellExtraArgs args(TRIGGERED_FULL_MASK);
+                        args.SpellValueOverrides.AddMod(SPELLVALUE_AURA_STACK, sSpellMgr->AssertSpellInfo(74396)->StackAmount);
+                        target->CastSpell(74396, SPELLVALUE_AURA_STACK, args);
```
whats going on here? 74396 being cast to WorldObject*?

Treeston replied:
typo on refactor

Shauren commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -855,17 +857,14 @@ void Spell::EffectTriggerSpell(SpellEffIndex effIndex)
-        values.AddSpellMod(SPELLVALUE_BASE_POINT1, damage);
-        values.AddSpellMod(SPELLVALUE_BASE_POINT2, damage);
-    }
+        for (int i = 0; i < 3; ++i)
```
use uint32 and MAX_EFFECTS instead of hardcoded 3

Shauren commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -902,18 +901,14 @@ void Spell::EffectTriggerMissileSpell(SpellEffIndex effIndex)
-        values.AddSpellMod(SPELLVALUE_BASE_POINT1, damage);
-        values.AddSpellMod(SPELLVALUE_BASE_POINT2, damage);
-    }
+        for (int i = 0; i < 3; ++i)
```
use uint32 and MAX_EFFECTS instead of hardcoded 3

Shauren commented on src/server/game/Spells/SpellEffects.cpp:
```diff
@@ -945,32 +940,28 @@ void Spell::EffectForceCast(SpellEffIndex effIndex)
-
-    SpellCastTargets targets;
-    targets.SetUnitTarget(m_caster);
+        for (int i = 0; i < 3; ++i)
```
use uint32 and MAX_EFFECTS instead of hardcoded 3

Shauren commented on src/server/scripts/EasternKingdoms/ZulAman/boss_hexlord.cpp:
```diff
@@ -1024,7 +1024,9 @@ class spell_hexlord_unstable_affliction : public SpellScriptLoader
             {
                 if (Unit* caster = GetCaster())
-                    caster->CastSpell(dispelInfo->GetDispeller(), SPELL_WL_UNSTABLE_AFFL_DISPEL, true, nullptr, GetEffect(EFFECT_0));
+                {
```
them brackets

Shauren commented on src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_halion.cpp:
```diff
@@ -348,7 +348,7 @@ class boss_halion : public CreatureScript
                             {
                                 _meteorStrikePos = target->GetPosition();
-                                me->CastSpell(_meteorStrikePos.GetPositionX(), _meteorStrikePos.GetPositionY(), _meteorStrikePos.GetPositionZ(), SPELL_METEOR_STRIKE, true, nullptr, nullptr, me->GetGUID());
+                                me->CastSpell({ _meteorStrikePos.GetPositionX(), _meteorStrikePos.GetPositionY(), _meteorStrikePos.GetPositionZ() }, SPELL_METEOR_STRIKE, me->GetGUID());
```
Probably could just use _meteorStrikePos directly?

Shauren commented on src/server/scripts/Northrend/ChamberOfAspects/RubySanctum/boss_halion.cpp:
```diff
@@ -1787,7 +1796,7 @@ class spell_halion_twilight_phasing : public SpellScriptLoader
             {
                 Unit* caster = GetCaster();
-                caster->CastSpell(caster->GetPositionX(), caster->GetPositionY(), caster->GetPositionZ(), SPELL_SUMMON_TWILIGHT_PORTAL, true);
+                caster->CastSpell({ caster->GetPositionX(), caster->GetPositionY(), caster->GetPositionZ() }, SPELL_SUMMON_TWILIGHT_PORTAL, true);
```
caster->GetPosition()

Treeston committed:
more cleanup

Shauren commented on src/server/scripts/Northrend/Ulduar/HallsOfLightning/boss_volkhan.cpp:
```diff
@@ -202,7 +202,7 @@ class boss_volkhan : public CreatureScript
 
                 // Why healing when just summoned?
-                summoned->CastSpell(summoned, SPELL_HEAT, false, nullptr, nullptr, me->GetGUID());
+                summoned->CastSpell(summoned, SPELL_HEAT, CastSpellExtraArgs().SetOriginalCaster(me->GetGUID()));
```
just me->GetGUID?

Treeston replied:
that would set triggered flag

Treeston replied:
(in almost all cases, casts that pass caster GUID are triggered, same for aura/item - i'm fairly confident that the ones that _don't_ are bugs, but didn't want to change them in the refactor)

Shauren commented on src/server/scripts/Outland/HellfireCitadel/ShatteredHalls/boss_nethekurse.cpp:
```diff
@@ -214,7 +214,7 @@ class boss_grand_warlock_nethekurse : public CreatureScript
                 //triggered spell of consumption does not properly show it's SpellVisual, wrong spellid?
                 summoned->CastSpell(summoned, SPELL_TEMPORARY_VISUAL, true);
-                summoned->CastSpell(summoned, SPELL_CONSUMPTION, false, 0, 0, me->GetGUID());
+                summoned->CastSpell(summoned, SPELL_CONSUMPTION, CastSpellExtraArgs().SetOriginalCaster(me->GetGUID()));
```
just guid?

Treeston replied:
same as above

Shauren commented on src/server/scripts/Spells/spell_dk.cpp:
```diff
@@ -1518,11 +1547,18 @@ class spell_dk_improved_unholy_presence : public SpellScriptLoader
+                    target->CastSpell(target, SPELL_DK_IMPROVED_UNHOLY_PRESENCE_TRIGGERED, args);
                 }
-
+                
```
whitespaaaaaaace

Shauren commented on src/server/scripts/Spells/spell_druid.cpp:
```diff
@@ -996,29 +1007,35 @@ class spell_dru_lifebloom : public SpellScriptLoader
             }
 
-            void AfterRemove(AuraEffect const* aurEff, AuraEffectHandleModes /*mode*/)
+            void _handleRemove(Unit* target, AuraEffect const* aurEff, uint32 stack)
```
Naming

Treeston committed:
cleanup

joschiwald commented on src/server/game/Entities/Unit/Unit.cpp:
```diff
@@ -958,133 +958,46 @@ void Unit::CastStop(uint32 except_spellid)
+            targets.SetGOTarget(goTarget);
+        else
+        {
+            TC_LOG_ERROR("entities.unit", "CastSpell: Invalid target with GUID %s passed to spell cast by %s\n", target->GetGUID().ToString().c_str(), GetGUID().ToString().c_str());
```
don't add the word `guid`, it is included in ObjectGuid::ToString
all log messages are finialized with \n by default, no need to add it here

Treeston committed:
minor text fix

Shauren commented:
Not approving the part where SPELLVALUE_MAX_TARGETS was moved to SpellMgr::LoadSpellInfoCorrections, rest is good

ariel- commented on src/server/game/Entities/Unit/Unit.h:
```diff
@@ -1278,20 +1287,10 @@ class TC_GAME_API Unit : public WorldObject
-        void CastCustomSpell(uint32 spellId, SpellValueMod mod, int32 value, Unit* victim = nullptr, TriggerCastFlags triggerFlags = TRIGGERED_NONE, Item* castItem = nullptr, AuraEffect const* triggeredByAura = nullptr, ObjectGuid originalCaster = ObjectGuid::Empty);
-        void CastCustomSpell(uint32 spellId, CustomSpellValues const& value, Unit* victim = nullptr, TriggerCastFlags triggerFlags = TRIGGERED_NONE, Item* castItem = nullptr, AuraEffect const* triggeredByAura = nullptr, ObjectGuid originalCaster = ObjectGuid::Empty);
+        // CastSpell's third arg can be a variety of things - check out CastSpellExtraArgs' constructors!
+        void CastSpell(SpellCastTargets const& targets, uint32 spellId, CastSpellExtraArgs const& args = {});
```
Can you default initialize a reference? or you need to instantiate a static empty instance to default to (like ObjectGuid::Empty)

Treeston replied:
You can default initialize a _const_ reference.

Shauren replied:
Yes, you can do that. It will get exactly the same behavior as ObjectGuid::Empty, residing somewhere in constant part of the executable but without a name to refer to in code

Treeston committed:
- split spell stuff from unit into new header

- the \n for josch
- move MAX_TARGETS back to script for shauren (and new ctor too)

joschiwald commented on src/server/game/Spells/Auras/SpellAuraDefines.h:
```diff
@@ -398,4 +398,40 @@ enum AuraObjectType
+// high byte (3 from 0..3) of UNIT_FIELD_BYTES_2
+enum ShapeshiftForm
+{
+    FORM_NONE = 0x00,
```
alignment

Faq commented on src/server/game/Spells/SpellMgr.cpp:
```diff
@@ -3276,17 +3253,15 @@ void SpellMgr::LoadSpellInfoCorrections()
-        28796, // Poison Bolt Volly - Faerlina
-        41303, // Soul Drain (Reliquary of Souls)
-        40243, // Crushing Shadows (Teron Gorefiend)
-        30657, // Quake (Magtheridon)
```
Why these were removed?

Treeston replied:
moved out to respective scripts, re @shuaren

Treeston committed:
alignment

joschiwald commented on src/server/game/Spells/SpellDefines.h:
```diff
@@ -0,0 +1,196 @@
+
+enum SpellModOp : uint8
+{
+    SPELLMOD_DAMAGE = 0,
```
alignment

Treeston committed:
more alignment
